R/lcd.R

```diff
@@ -95,3 +104,277 @@ lcd_base <- function() {
+  MonthlyGreatestPrecip="numeric",
+  MonthlyGreatestPrecipDate="character",
+  MonthlyGreatestSnowDepth="numeric",
+  MonthlyGreatestSnowDepthDate="character",
+  MonthlyGreatestSnowfall="numeric",
+  MonthlyGreatestSnowfallDate="character",
+  MonthlyMaxSeaLevelPressureValue="numeric",
+  MonthlyMaxSeaLevelPressureValueDate="character",
+  MonthlyMaxSeaLevelPressureValueTime="character",
+  MonthlyMaximumTemperature="numeric",
+  MonthlyMeanTemperature="numeric",
+  MonthlyMinSeaLevelPressureValue="numeric",
+  MonthlyMinSeaLevelPressureValueDate="character",
+  MonthlyMinSeaLevelPressureValueTime="character",
+  MonthlyMinimumTemperature="numeric",
+  MonthlySeaLevelPressure="numeric",
+  MonthlyStationPressure="numeric",
+  MonthlyTotalLiquidPrecipitation="numeric",
+  MonthlyTotalSnowfall="numeric",
+  MonthlyWetBulb="numeric",
+  AWND="numeric",
+  CDSD="numeric",
+  CLDD="numeric",
+  DSNW="numeric",
+  HDSD="numeric",
+  HTDD="numeric",
+  NormalsCoolingDegreeDay="numeric",
+  NormalsHeatingDegreeDay="numeric",
+  ShortDurationEndDate005="character",
+  ShortDurationEndDate010="character",
+  ShortDurationEndDate015="character",
+  ShortDurationEndDate020="character",
+  ShortDurationEndDate030="character",
+  ShortDurationEndDate045="character",
+  ShortDurationEndDate060="character",
+  ShortDurationEndDate080="character",
+  ShortDurationEndDate100="character",
+  ShortDurationEndDate120="character",
+  ShortDurationEndDate150="character",
+  ShortDurationEndDate180="character",
+  ShortDurationPrecipitationValue005="numeric",
+  ShortDurationPrecipitationValue010="numeric",
+  ShortDurationPrecipitationValue015="numeric",
+  ShortDurationPrecipitationValue020="numeric",
+  ShortDurationPrecipitationValue030="numeric",
+  ShortDurationPrecipitationValue045="numeric",
+  ShortDurationPrecipitationValue060="numeric",
+  ShortDurationPrecipitationValue080="numeric",
+  ShortDurationPrecipitationValue100="numeric",
+  ShortDurationPrecipitationValue120="numeric",
+  ShortDurationPrecipitationValue150="numeric",
+  ShortDurationPrecipitationValue180="numeric",
+  REM="character",
+  BackupDirection="character",
+  BackupDistance="character",
+  BackupDistanceUnit="character",
+  BackupElements="character",
+  BackupElevation="character",
+  BackupEquipment="character",
+  BackupLatitude="character",
+  BackupLongitude="character",
+  BackupName="character",
+  WindEquipmentChangeDate="character") {
+  # setting the column types for data as described in:
+  # https://www1.ncdc.noaa.gov/pub/data/ish/ish-format-document.pdf
+  # and: https://www.ncei.noaa.gov/data/local-climatological-data/doc/LCD_documentation.pdf
+  # daily and monthly values might include special indicators with the value or
+  # instead of a value
+  # s = suspect value, T = Trace, M = missing, * extreme max or min for month
+  # to retain this information, values are kept as character for user to handle
+  argg <- unlist(c(as.list(environment())))
+  return(argg)
```

Comment:
seems like we should check here that the input values to each parameter is a proper R class, e.g., `lcd_columns(STATION = 5)` doesn't make much sense. although, fread does give a nice error message at least when you use it in `lcd`

```r
lcd(station = "74746003904", year = 2010, col_types = lcd_columns(STATION = 5))
#> Error in safe_read_csv(path, col_types = col_types) :
#>   Column 'STATION' was requested to be '5' but fread encountered the following error:
#> 	no method or default for coercing “character” to “5”
#> so the column has been left as type 'character'
```

Reply:
Agreed. I've added a function to zzz.r that returns an error message if wrong types are entered. This is run by both `lcd()` and `lcd_columns()` with some logic to reduce the chance it is run twice.

```r
## works
x.1 <- lcd(station = "74746003904", year = 2010, col_types = c(STATION="character"))
x.1[c(1:6)]

# A tibble: 11,154 x 6
   station     date                latitude longitude elevation name                                   
   <chr>       <dttm>                 <dbl>     <dbl>     <dbl> <chr>                                  
 1 74746003904 2010-01-01 00:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 2 74746003904 2010-01-01 01:24:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 3 74746003904 2010-01-01 01:51:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 4 74746003904 2010-01-01 01:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 5 74746003904 2010-01-01 02:35:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 6 74746003904 2010-01-01 02:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 7 74746003904 2010-01-01 03:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 8 74746003904 2010-01-01 04:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 9 74746003904 2010-01-01 05:27:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
10 74746003904 2010-01-01 05:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
# ... with 11,144 more rows
```

```r 
## works
x.2 <- lcd(station = "74746003904", year = 2010, col_types = lcd_columns(DATE = "character"))
x.2[c(1:6)]
# A tibble: 11,154 x 6
   station     date                latitude longitude elevation name                                   
   <chr>       <chr>                  <dbl>     <dbl>     <dbl> <chr>                                  
 1 74746003904 2010-01-01T00:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 2 74746003904 2010-01-01T01:24:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 3 74746003904 2010-01-01T01:51:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 4 74746003904 2010-01-01T01:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 5 74746003904 2010-01-01T02:35:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 6 74746003904 2010-01-01T02:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 7 74746003904 2010-01-01T03:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 8 74746003904 2010-01-01T04:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 9 74746003904 2010-01-01T05:27:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
10 74746003904 2010-01-01T05:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
# ... with 11,144 more rows
```

```r
## expect errors
col_types <- c(STATION=5)
lcd(station = "74746003904", year = 2010, col_types = col_types)

Error in lcd(station = "74746003904", year = 2010, col_types = col_types) : 
  STATION must equal a valid R class ('character', 'integer', 'numeric', 'factor', 'integer64', 'POSIXct')
```

``r
## expect errors
lcd_columns(STATION = 5)

Error in lcd_columns(STATION = 5) : 
  STATION must equal a valid R class ('character', 'integer', 'numeric', 'factor', 'integer64', 'POSIXct')
```

Hopefully these are useful error messages for users.

Reply:
Looks great